src/Compilers/CSharp/Portable/Lowering/LocalRewriter/DelegateCacheContainer.cs

```diff
@@ -0,0 +1,77 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.Symbols;
+
+/// <summary>
+/// This type is synthesized to hold the cached delegates that target static method groups.
+/// </summary>
+internal sealed class DelegateCacheContainer : SynthesizedContainer
+{
+    private readonly Symbol _containingSymbol;
+    private readonly NamedTypeSymbol? _constructedContainer;
+    private readonly Dictionary<(TypeSymbol, MethodSymbol), FieldSymbol> _delegateFields = new();
+
+    /// <summary>Creates a type scoped concrete delegate cache container.</summary>
+    internal DelegateCacheContainer(TypeSymbol containingType, int generationOrdinal)
+        : base(GeneratedNames.DelegateCacheContainerType(generationOrdinal), containingMethod: null)
+    {
+        Debug.Assert(containingType.IsDefinition);
+        _containingSymbol = containingType;
+    }
+
+    /// <summary>Creates a method scoped generic delegate cache container.</summary>
+    internal DelegateCacheContainer(MethodSymbol containingMethod, int topLevelMethodOrdinal, int ownerUniqueId, int generationOrdinal)
+        : base(GeneratedNames.DelegateCacheContainerType(generationOrdinal, containingMethod.Name, topLevelMethodOrdinal, ownerUniqueId), containingMethod)
+    {
+        Debug.Assert(containingMethod.IsDefinition);
+        _containingSymbol = containingMethod.ContainingType;
+        _constructedContainer = Construct(ConstructedFromTypeParameters);
```

Comment:
>Construct [](http://example.com/codeflow?start=32&length=9)

If this type is not generic. Isn't this call going to throw? What am I missing?

Reply:
Let me add an assert.